Failed to update tomcat: DependencyInfo: cannot deserialize from Object value

Now that #557 is fixed, I get this error:
```
Start: Install tomcat
Step 'Install or update software' ended with failure.
Step 'ide (update)' failed: java.lang.IllegalStateException: Failed to load D:\projects\_ide\urls\tomcat\tomcat\dependencies.json
 Step 'Install or update software' failed: null
  Step 'Install tomcat (*)' failed: java.lang.IllegalStateException: Failed to load D:\projects\_ide\urls\tomcat\tomcat\dependencies.json
3 step(s) failed out of 12 steps.
An unexpected error occurred!
We are sorry for the inconvenience.
Please check the error below, resolve it and try again.
If the error is not on your end (network connectivity, lack of permissions, etc.) please file a bug:
https://github.com/devonfw/IDEasy/issues/new?assignees=&labels=bug&projects=&template=bug.md&title=IllegalStateException%3A+Failed+to+load+D%3A%5Cprojects%5C_ide%5Curls%5Ctomcat%5Ctomcat%5Cdependencies.json
java.lang.IllegalStateException: Failed to load D:\projects\_ide\urls\tomcat\tomcat\dependencies.json
        at com.devonfw.tools.ide.url.model.file.UrlDependencyFile.doLoad(UrlDependencyFile.java:85)
        at com.devonfw.tools.ide.url.model.file.AbstractUrlFile.load(AbstractUrlFile.java:39)
        at com.devonfw.tools.ide.url.model.folder.UrlEdition.getDependencyFile(UrlEdition.java:60)
        at com.devonfw.tools.ide.tool.Dependency.readJson(Dependency.java:53)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.readDependencies(LocalToolCommandlet.java:375)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.installTool(LocalToolCommandlet.java:193)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.installTool(LocalToolCommandlet.java:177)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.installTool(LocalToolCommandlet.java:151)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.doInstall(LocalToolCommandlet.java:78)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.doInstall(LocalToolCommandlet.java:65)
        at com.devonfw.tools.ide.tool.ToolCommandlet.install(ToolCommandlet.java:233)
        at com.devonfw.tools.ide.commandlet.AbstractUpdateCommandlet.updateSoftware(AbstractUpdateCommandlet.java:170)
        at com.devonfw.tools.ide.commandlet.AbstractUpdateCommandlet.run(AbstractUpdateCommandlet.java:52)
        at com.devonfw.tools.ide.commandlet.UpdateCommandlet.run(UpdateCommandlet.java:29)
        at com.devonfw.tools.ide.context.AbstractIdeContext.applyAndRun(AbstractIdeContext.java:856)
        at com.devonfw.tools.ide.context.AbstractIdeContext.run(AbstractIdeContext.java:765)
        at com.devonfw.tools.ide.cli.Ideasy.runOrThrow(Ideasy.java:89)
        at com.devonfw.tools.ide.cli.Ideasy.run(Ideasy.java:52)
        at com.devonfw.tools.ide.cli.Ideasy.main(Ideasy.java:29)
        at java.base@21.0.2/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH)
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `com.devonfw.tools.ide.url.model.file.json.DependencyInfo`: cannot deserialize from Object value (no delegate- or property-based Creator): this appears to be a native image, in which case you may need to configure reflection for the class that is to be deserialized
 at [Source: (BufferedReader); line: 4, column: 7] (through reference chain: java.util.HashMap["[10.1,11.0.0)"]->java.util.ArrayList[0])
        at com.fasterxml.jackson.databind.DeserializationContext.reportBadDefinition(DeserializationContext.java:1909)
        at com.fasterxml.jackson.databind.DatabindContext.reportBadDefinition(DatabindContext.java:408)
        at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1349)
        at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1417)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:352)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:185)
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeFromArray(CollectionDeserializer.java:359)
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:244)
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:28)
        at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBind(MapDeserializer.java:557)
        at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:451)
        at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:32)
        at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323)
        at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4730)
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3698)
        at com.devonfw.tools.ide.url.model.file.UrlDependencyFile.doLoad(UrlDependencyFile.java:83)
        ... 19 more


Error: IDEasy failed with exit code 255
```

We really need to test builds of GraalVM in our Github Actions if the behaviour of JVM and GraalVM native image differs significantly.

Retested after merging #636:
```
java.lang.IllegalStateException: Failed to load D:\projects\_ide\urls\tomcat\tomcat\dependencies.json
        at com.devonfw.tools.ide.url.model.file.UrlDependencyFile.doLoad(UrlDependencyFile.java:85)
        at com.devonfw.tools.ide.url.model.file.AbstractUrlFile.load(AbstractUrlFile.java:39)
        at com.devonfw.tools.ide.url.model.folder.UrlEdition.getDependencyFile(UrlEdition.java:60)
        at com.devonfw.tools.ide.tool.Dependency.readJson(Dependency.java:53)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.readDependencies(LocalToolCommandlet.java:375)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.installTool(LocalToolCommandlet.java:193)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.installTool(LocalToolCommandlet.java:177)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.installTool(LocalToolCommandlet.java:151)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.doInstall(LocalToolCommandlet.java:78)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.doInstall(LocalToolCommandlet.java:65)
        at com.devonfw.tools.ide.tool.ToolCommandlet.install(ToolCommandlet.java:231)
        at com.devonfw.tools.ide.commandlet.AbstractUpdateCommandlet.updateSoftware(AbstractUpdateCommandlet.java:170)
        at com.devonfw.tools.ide.commandlet.AbstractUpdateCommandlet.run(AbstractUpdateCommandlet.java:52)
        at com.devonfw.tools.ide.commandlet.UpdateCommandlet.run(UpdateCommandlet.java:29)
        at com.devonfw.tools.ide.context.AbstractIdeContext.applyAndRun(AbstractIdeContext.java:856)
        at com.devonfw.tools.ide.context.AbstractIdeContext.run(AbstractIdeContext.java:765)
        at com.devonfw.tools.ide.cli.Ideasy.runOrThrow(Ideasy.java:89)
        at com.devonfw.tools.ide.cli.Ideasy.run(Ideasy.java:52)
        at com.devonfw.tools.ide.cli.Ideasy.main(Ideasy.java:29)
        at java.base@21.0.2/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH)
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `com.devonfw.tools.ide.url.model.file.json.DependencyInfo`: cannot deserialize from Object value (no delegate- or property-based Creator): this appears to be a native image, in which case you may need to configure reflection for the class that is to be deserialized
 at [Source: (BufferedReader); line: 4, column: 7] (through reference chain: java.util.HashMap["[10.1,11.0.0)"]->java.util.ArrayList[0])
        at com.fasterxml.jackson.databind.DeserializationContext.reportBadDefinition(DeserializationContext.java:1909)
        at com.fasterxml.jackson.databind.DatabindContext.reportBadDefinition(DatabindContext.java:408)
        at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1349)
        at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1417)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:352)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:185)
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeFromArray(CollectionDeserializer.java:359)
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:244)
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:28)
        at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBind(MapDeserializer.java:557)
        at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:451)
        at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:32)
        at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323)
        at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4730)
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3698)
        at com.devonfw.tools.ide.url.model.file.UrlDependencyFile.doLoad(UrlDependencyFile.java:83)
        ... 19 more
```

--- a/cli/src/main/java/com/devonfw/tools/ide/json/DependencyInfoDeserializer.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/json/DependencyInfoDeserializer.java
@@ -1,0 +1,49 @@
+package com.devonfw.tools.ide.json;
+
+import java.io.IOException;
+
+import com.devonfw.tools.ide.url.model.file.json.DependencyInfo;
+import com.devonfw.tools.ide.version.VersionRange;
+import com.fasterxml.jackson.core.JacksonException;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonToken;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JsonDeserializer;
+
+/**
+ * {@link JsonDeserializer} for {@link DependencyInfo}.
+ */
+public class DependencyInfoDeserializer extends JsonDeserializer<DependencyInfo> {
+
+  @Override
+  public DependencyInfo deserialize(JsonParser p, DeserializationContext context) throws IOException, JacksonException {
+
+    JsonToken token = p.getCurrentToken();
+    if (token == JsonToken.START_OBJECT) {
+      token = p.nextToken();
+      String tool = null;
+      VersionRange versionRange = null;
+      while (token == JsonToken.FIELD_NAME) {
+        String property = p.currentName();
+        if (property.equals("tool")) {
+          token = p.nextToken();
+          assert token == JsonToken.VALUE_STRING;
+          tool = p.getValueAsString();
+        } else if (property.equals("versionRange")) {
+          token = p.nextToken();
+          assert (token == JsonToken.START_OBJECT) || (token == JsonToken.VALUE_STRING);
+          versionRange = p.readValueAs(VersionRange.class);
+        } else {
+          // ignore unknown property
+          // currently cannot log here due to https://github.com/devonfw/IDEasy/issues/404
+        }
+        token = p.nextToken();
+      }
+      if ((tool != null) && (versionRange != null)) {
+        return new DependencyInfo(tool, versionRange);
+      }
+    }
+    throw new IllegalArgumentException("Invalid JSON for DependencyInfo!");
+  }
+
+}

--- a/cli/src/main/java/com/devonfw/tools/ide/json/JsonMapping.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/json/JsonMapping.java
@@ -1,5 +1,6 @@
 package com.devonfw.tools.ide.json;
 
+import com.devonfw.tools.ide.url.model.file.json.DependencyInfo;
 import com.devonfw.tools.ide.version.VersionIdentifier;
 import com.devonfw.tools.ide.version.VersionRange;
 import com.fasterxml.jackson.annotation.JsonInclude;
@@ -28,6 +29,7 @@ public class JsonMapping {
     SimpleModule customModule = new SimpleModule();
     customModule.addDeserializer(VersionIdentifier.class, new VersionIdentifierDeserializer());
     customModule.addDeserializer(VersionRange.class, new VersionRangeDeserializer());
+    customModule.addDeserializer(DependencyInfo.class, new DependencyInfoDeserializer());
     customModule.addKeyDeserializer(VersionRange.class, new VersionRangeKeyDeserializer());
     mapper = mapper.registerModule(customModule);
     return mapper;

--- a/cli/src/test/java/com/devonfw/tools/ide/json/JsonMappingTest.java
+++ b/cli/src/test/java/com/devonfw/tools/ide/json/JsonMappingTest.java
@@ -5,6 +5,10 @@ import java.time.Instant;
 import org.assertj.core.api.Assertions;
 import org.junit.jupiter.api.Test;
 
+import com.devonfw.tools.ide.url.model.file.json.DependencyInfo;
+import com.devonfw.tools.ide.version.BoundaryType;
+import com.devonfw.tools.ide.version.VersionIdentifier;
+import com.devonfw.tools.ide.version.VersionRange;
 import com.fasterxml.jackson.databind.ObjectMapper;
 
 /**
@@ -26,12 +30,12 @@ public class JsonMappingTest extends Assertions {
   @Test
   public void testReadInstant() throws Exception {
 
-    // given
+    // arrange
     String value = INSTANT_VALUE_JSON;
-    // when
+    // act
     ObjectMapper mapper = JsonMapping.create();
     Instant instant = mapper.readValue(value, Instant.class);
-    // then
+    // assert
     assertThat(instant).isEqualTo(INSTANT_VALUE);
   }
 
@@ -43,13 +47,27 @@ public class JsonMappingTest extends Assertions {
   @Test
   public void testWriteInstant() throws Exception {
 
-    // given
+    // arrange
     Instant value = INSTANT_VALUE;
-    // when
+    // act
     ObjectMapper mapper = JsonMapping.create();
     String json = mapper.writeValueAsString(value);
-    // then
+    // assert
     assertThat(json).isEqualTo(INSTANT_VALUE_JSON);
+  }
+
+  @Test
+  public void testReadDependencies() throws Exception {
+
+    // arrange
+    String json = "{\"tool\": \"java\",\"versionRange\": \"[11,21_35]\"}";
+    VersionRange expectedVersionRange = VersionRange.of(VersionIdentifier.of("11"), VersionIdentifier.of("21_35"), BoundaryType.CLOSED);
+    // act
+    ObjectMapper mapper = JsonMapping.create();
+    DependencyInfo dependencyInfo = mapper.readValue(json, DependencyInfo.class);
+    // assert
+    assertThat(dependencyInfo.tool()).isEqualTo("java");
+    assertThat(dependencyInfo.versionRange()).isEqualTo(expectedVersionRange);
   }
 
 }

